File: 0341-flatten-nested-list-iterator/0341-flatten-nested-list-iterator.py
# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
#class NestedInteger:
#    def isInteger(self) -> bool:
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        """
#
#    def getInteger(self) -> int:
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        """
#
#    def getList(self) -> [NestedInteger]:
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        """

class NestedIterator:
    def __init__(self, nestedList: [NestedInteger]):
        # SC- O(n)
        
        self.stack = []
        # push in reverse so that top of stack has first element
        for ni in reversed(nestedList):
            self.stack.append(ni)
    
    def next(self) -> int:
        # TC- O(1)
        # hasNext ensures top of stack is always an integer
        return self.stack.pop().getInteger()
        
    def hasNext(self) -> bool:
        # if stack is empty, return False. Else, keep iterating until top of stack is integer and return true.
        # TC- amortized O(1)
        
        while self.stack:
            top = self.stack[-1]
            if top.isInteger():
                return True
            
            self.stack.pop()
            for ni in reversed(top.getList()):
                self.stack.append(ni)
                
        return False

# Your NestedIterator object will be instantiated and called as such:
# i, v = NestedIterator(nestedList), []
# while i.hasNext(): v.append(i.next())

File: 0341-flatten-nested-list-iterator/0341-flatten-nested-list-iterator.cpp
/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * class NestedInteger {
 *   public:
 *     // Return true if this NestedInteger holds a single integer, rather than a nested list.
 *     bool isInteger() const;
 *
 *     // Return the single integer that this NestedInteger holds, if it holds a single integer
 *     // The result is undefined if this NestedInteger holds a nested list
 *     int getInteger() const;
 *
 *     // Return the nested list that this NestedInteger holds, if it holds a nested list
 *     // The result is undefined if this NestedInteger holds a single integer
 *     const vector<NestedInteger> &getList() const;
 * };
 */

class NestedIterator {
private:
    // SC- O(n)
    stack<NestedInteger> st;
public:
    NestedIterator(vector<NestedInteger> &nestedList) {
        // TC- O(n)
        for(int j=nestedList.size()-1; j>=0; j--)
            st.push(nestedList[j]);
    }
    
    int next() {
        // TC- O(1)
        int val = st.top().getInteger();
        st.pop();
        return val;
    }
    
    bool hasNext() {
        // TC- amortized O(1)
        while(!st.empty()){
            if (st.top().isInteger())
                return true;
            
            auto l = st.top().getList();
            st.pop();
            for(int i=l.size()-1; i>=0; i--)
                st.push(l[i]);
        }

        return false;
    }
};

/**
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i(nestedList);
 * while (i.hasNext()) cout << i.next();
 */